math/src/polynom/multi_linear.rs

```diff
@@ -0,0 +1,244 @@
+        for chunk in self.evaluations.chunks_exact(2) {
+            p0.push(chunk[0]);
+            p1.push(chunk[1]);
+        }
+
+        (MultiLinearPoly::from_evaluations(p0), MultiLinearPoly::from_evaluations(p1))
+    }
+}
+
+impl<E: FieldElement> Index<usize> for MultiLinearPoly<E> {
+    type Output = E;
+
+    fn index(&self, index: usize) -> &E {
+        &(self.evaluations[index])
+    }
+}
+
+impl<E> Serializable for MultiLinearPoly<E>
+where
+    E: FieldElement,
+{
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        let Self { evaluations } = self;
+        evaluations.write_into(target);
+    }
+}
+
+impl<E> Deserializable for MultiLinearPoly<E>
+where
+    E: FieldElement,
+{
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        Ok(Self {
+            evaluations: Deserializable::read_from(source)?,
+        })
+    }
+}
+
+// EQ FUNCTION
+// ================================================================================================
+
+/// The EQ (equality) function is the binary function defined by
+///
+/// $$
+/// EQ:    {0 , 1}^ν ⛌ {0 , 1}^ν ⇾ {0 , 1}
+///   ((x_0, ..., x_{ν - 1}), (y_0, ..., y_{ν - 1})) ↦ \prod_{i = 0}^{ν - 1} (x_i * y_i + (1 - x_i)
+/// * (1 - y_i))
+/// $$
+///
+/// Taking its multi-linear extension $EQ^{~}$, we can define a basis for the set of multi-linear
+/// polynomials in ν variables by
+///         $${EQ^{~}(., (y_0, ..., y_{ν - 1})): (y_0, ..., y_{ν - 1}) ∈ {0 , 1}^ν}$$
+/// where each basis function is a function of its first argument. This is called the Lagrange or
+/// evaluation basis for evaluation set ${0 , 1}^ν$.
+///
+/// Given a function $(f: {0 , 1}^ν ⇾ 𝔽)$, its multi-linear extension (i.e., the unique
+/// mult-linear polynomial extending `f` to $(f^{~}: 𝔽^ν ⇾ 𝔽)$ and agreeing with it on ${0 , 1}^ν$) is
+/// defined as the summation of the evaluations of f against the Lagrange basis.
+/// More specifically, given $(r_0, ..., r_{ν - 1}) ∈ 𝔽^ν$, then:
+///
+/// $$
+///     f^{~}(r_0, ..., r_{ν - 1}) = \sum_{(y_0, ..., y_{ν - 1}) ∈ {0 , 1}^ν}
+///                  f(y_0, ..., y_{ν - 1}) EQ^{~}((r_0, ..., r_{ν - 1}), (y_0, ..., y_{ν - 1}))
+/// $$
+///
+/// We call the Lagrange kernel the evaluation of the EQ^{~} function at
+/// $((r_0, ..., r_{ν - 1}), (y_0, ..., y_{ν - 1}))$ for all $(y_0, ..., y_{ν - 1}) ∈ {0 , 1}^ν$ for
+/// a fixed $(r_0, ..., r_{ν - 1}) ∈ 𝔽^ν$.
+///
+/// [`EqFunction`] represents EQ^{~} the multi-linear extension of
+///
+/// $((y_0, ..., y_{ν - 1}) ↦ EQ((r_0, ..., r_{ν - 1}), (y_0, ..., y_{ν - 1})))$
+///
+/// and contains a method to generate the Lagrange kernel for defining evaluations of multi-linear
+/// extensions of arbitrary functions $(f: {0 , 1}^ν ⇾ 𝔽)$ at a given point $(r_0, ..., r_{ν - 1})$
+/// as well as a method to evaluate $EQ^{~}((r_0, ..., r_{ν - 1}), (t_0, ..., t_{ν - 1})))$ for
+/// $(t_0, ..., t_{ν - 1}) ∈ 𝔽^ν$.
+pub struct EqFunction<E> {
+    r: Vec<E>,
+}
```

Comment:
Is `r` expected to be a relatively small vector here? If so, maybe worth using `SmallVec` here as well.

Reply:
Changed